test: Add UI integration tests

cryptotavares opened the pull request:
## **Description**

This PR introduces UI integration tests to the Extension codebase.

By introducing UI Integration tests we are not just providing a new testing tool but also aiming to better clarify the differences between unit, integration and e2e tests, as well as when developers should aim at each.


### **What are integration tests?**
Integration tests evaluate the interaction between multiple components within the application, ensuring they work together as intended. For MetaMask Extension, we are aiming to make UI integration tests particularly focused on page-level components and are conducted in the context of the full UI application.

**When to write integration tests?**
Integration tests should be written for page-level components. These tests should simulate user journeys and validate the application's behavior in a full app context.

**Generic guidelines**
* **Full App Context**: Always test page-level components in the context of the full application. This approach ensures that tests reflect real user scenarios and interactions within the app.
* **Focus on User Journeys**: Design tests to mimic actual user actions and flows. This helps in identifying issues that could affect the user experience.
* **Avoid Testing Non-Page Components**: Integration tests should not be written for components other than page-level components. Other components should be tested using unit tests.
* **Avoid Snapshots**: Refrain from using snapshot testing in integration tests. They tend to be less meaningful in the context of full app testing and can lead to brittle tests.
* **Test Location**: Place integration test files alongside the page component they are testing, with an appropriate naming convention (e.g., page.integration.test.ts).
* **No Mocks**: Keep mocking to minimum. Ideally only the background connection, or any network request (fired from the UI) should be mocked.

### **What are unit tests?**
Unit tests understanding is a bit more diffuse. As there is no consensus on how they should be written and what they should test. But generically, [as captured by Martin Fowler](https://martinfowler.com/articles/2021-test-shapes.html), there's two lines of thought: solitary unit tests and sociable unit tests. Within MetaMask there are multiple examples from tests and discussions highlighting the two types (for example you can check the discussion on [this PR](https://www.google.com/url?q=https://github.com/MetaMask/core/pull/3827%23discussion_r1469377179&sa=D&source=editors&ust=1716475694980436&usg=AOvVaw3oTcfVgfRuiQFSDChZPrFM))
Unit tests examine individual components or functions in isolation or within their immediate context. At MetaMask we encourage a flexible approach to unit testing, recognizing the spectrum between sociable and solitary unit tests.

**When to write unit tests?**
All components, except page-level components, should be validated through unit tests. These tests focus on the component's internal logic and functionality. Integration tests are preferred for page-level components. However, developers may choose to write sociable unit tests for specific scenarios where integration tests might not be necessary or practical.

**Generic guidelines**
* **Flexibility in Isolation**: Depending on the test's purpose, developers can choose between sociable unit tests (without mocking child components) and solitary unit tests (with mocked dependencies) to best suit the testing needs.
* **Colocation**: Keep unit test files next to their corresponding implementation files. This practice enhances test discoverability and maintenance.
* **Granularity and Focus**: Ensure unit tests are focused and granular, targeting specific functionalities. The choice between sociable and solitary testing should be guided by the test's objectives and the component's role within the application.
* **Unit Tests for Page Components**: While integration tests are the primary method for testing page-level components, developers have the discretion to use unit tests for specific cases. This flexibility allows for targeted testing of component logic or behavior that may not require the full app context.

### **What are e2e tests?**
e2e tests simulate real user scenarios from start to finish, testing the application as a whole. In the Extension, e2e tests are tests that strive to test a real extension app (including background, contentscript and ui processes) in a controlled environment from a user perspective.

**When to write e2e tests?**
Testing application's integration with external systems or services through critical user journeys. This includes any interactions between the UI, background process, dapp, blockchain, etc.

**Generic guidelines**
* **Realistic Scenarios**: Test scenarios that closely mimic real user actions and flows.

### **Broader guidelines**
* **Code fencing:** is very problematic for unit/integration testing. We should avoid it.
* **jest manual mocks:** This types of mocks are automatically picked up by jest for all tests. We should be very careful when writing this types of mocks as they will be shared across all tests. And with integration tests, as we are aiming to mock as little as possible, we should avoid this at all costs. 

### **Conclusion**
To sum up, understanding the distinction between integration and unit tests and applying them appropriately is crucial for maintaining MetaMask's code quality and reliability. Integration tests offer a broad view of user interactions and system integration for page-level components, while unit tests provide detailed insights into the functionality of individual components. By adhering to these guidelines and embracing the flexibility between sociable and solitary unit tests, developers can effectively contribute to MetaMask's robustness and user satisfaction.

### **References**
Some of the references used to sustain the UI integration tests approach are:
https://martinfowler.com/articles/2021-test-shapes.html
https://martinfowler.com/bliki/TestPyramid.html
https://martinfowler.com/bliki/IntegrationTest.html
https://kentcdodds.com/blog/write-tests#how-to-write-more-integration-tests
https://kentcdodds.com/blog/static-vs-unit-vs-integration-vs-e2e-tests
https://redux.js.org/usage/writing-tests#guiding-principles
https://testing-library.com/docs/guiding-principles

## Relevant changes
* UI integration tests configuration: jest config file, specific setup handlers, package.json scripts
* Test renderer helper for integration tests: split redux store initialisation from render the application; add helper to render the actual UI application given a metamask state (which is just a flatten controllers state served by the background process)
* Remove react router dom manual mock: manual mocks are picked up by jest, and this was preventing proper routing for the integration tests. Removed the generic manual mock and added specific mocks to all unit tests that required it.
* Two UI integration tests implemented (derived from existing e2e tests): personal sign confirmation (new design) and wallet created events
* Added CI job to run UI integration tests
* Added vscode launch.json config to attach debugger to UI integration tests

## Codespaces
[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24428?quickstart=1)

## **Manual testing steps**
In the terminal run:
```sh
yarn test:integration
```


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

cryptotavares committed:
test: initial integration tests

cryptotavares committed:
fix: broken unit tests

Unit tests broke aftre removing
react-router-dom manual mocks.
This commit adds the mocking to each
individual test file, preventing the
manual mocks to interfere with the
integration tests

cryptotavares committed:
fix: jest config coverage path

cryptotavares committed:
chore: add missing prop to init state

cryptotavares committed:
fix: prettier

cryptotavares committed:
fix: unit tests

cryptotavares committed:
fix: integration test update signature title desc

cryptotavares committed:
test: add wallet created events integration test

pnarayanaswamy committed:
chore: add github action to run integration tests

cryptotavares committed:
fix: lint

cryptotavares committed:
chore: move integration test data to correct dir

cryptotavares committed:
fix: unit tests

cryptotavares committed:
chore: add integration tests debugger

cryptotavares committed:
chore: shim fetch to handle relative urls

browser fetch can handle relative urls to fetch
local files. We are leveraging that for translation
files. By shimming the fetch for jest test env
we are able to mimic the same behaviour, thus no
needing to manually pass the translations
to the i18n provider

cryptotavares committed:
chore: update integration test renderer helper

As fetch is now shimmed and supports
getting the translation files, we can now
just render the Root component for UI
integration tests

cryptotavares commented on test/helpers/setup-helper.js:
```diff
@@ -89,10 +91,35 @@ window.document.body.appendChild(popoverContent);
 const { default: fetch, Headers, Request, Response } = require('node-fetch');
 
-Object.assign(window, { fetch, Headers, Request, Response });
+const shimmedFetch = (url, ...args) => {
```
We need to shim fetch in order to support relative URLs fetch (which in a node environment translates to accessing filesystem files). This is used by the i18n provider to get the correct locale translation file (messages.json).

matthewwalsh0 commented on .circleci/config.yml:
```diff
@@ -1747,6 +1752,23 @@ jobs:
       - store_test_results:
           path: test/test-results/junit.xml
 
+  test-integration-main:
```
Minor, but if there is no alternate types of integration tests (like the development unit tests), could we just call this `test-integration`?

matthewwalsh0 commented on jest.integration.config.js:
```diff
@@ -0,0 +1,37 @@
+    '<rootDir>/shared/**/*.(js|ts|tsx)',
+    '<rootDir>/ui/**/*.(js|ts|tsx)',
+  ],
+  coverageDirectory: './coverage/integration',
```
Are we concerned about coverage with integration tests?

My understanding was we're just using them to supplement E2E coverage, or are we actively trying to cover all pages?

cryptotavares replied:
Yes I don't think that test coverage will add a ton of value here. The real value would be on use case coverage but there's no easy way to measure that. The though was, well better to had the coverage config for the code coverage that devs are used, than nothing. (I this coverage report is not being merged with any of the unit tests coverage reports neither being reported to any tool - like codecov or sonarcloud... for now it is just for local generation and support the developer adding integration tests)

Gudahtt replied:
I would expect coverage to be useful here. We're currently striving to increase coverage as much as possible, with the goal of getting to 100%. But certain components (particularly page-level components) would be pointless to test in an isolationist way by mocking everything. The only useful way of testing it would be with an integration test. Considering integration test coverage would be one way of reducing pressure to write poor unit tests.

But we can figure this out in a later PR.

matthewwalsh0 commented on test/e2e/tests/confirmations/header.spec.js:
Are these relevant to the PR, I can see some UI changes also?

matthewwalsh0 replied:
My understanding is this is due to the additions of `data-testid` to support the example integration tests.

cryptotavares replied:
Yes it was just to add `data-testid` to help create the personalSign test (we could revert that from this PR.. but honestly feels like a small win for the personalSign example test added)

matthewwalsh0 commented on test/helpers/setup-helper.js:
```diff
@@ -89,10 +91,35 @@ window.document.body.appendChild(popoverContent);
-Object.assign(window, { fetch, Headers, Request, Response });
+const shimmedFetch = (url, ...args) => {
+  return fetch(url, ...args).catch((error) => {
+    if (error.message === 'Only absolute URLs are supported') {
```
For readability, should we move this block into a separate function like `handleRelativeFetchPaths`?

We could even import this shim from a separate file given the amount of logic involved?

matthewwalsh0 replied:
Could we invert this check and throw here to avoid the below nesting?

matthewwalsh0 commented on test/helpers/setup-helper.js:
```diff
@@ -89,10 +91,35 @@ window.document.body.appendChild(popoverContent);
+const shimmedFetch = (url, ...args) => {
+  return fetch(url, ...args).catch((error) => {
+    if (error.message === 'Only absolute URLs are supported') {
+      const regex = /_locales\/([^/]+)\/messages\.json/gu;
```
Do we need to limit it to the messages, or could we keep it general to any relative URL?

cryptotavares replied:
We do not... but for each available static asset, the path for it will depend on the static assets generated/placed during build ([here](https://github.com/MetaMask/metamask-extension/blob/develop/development/build/static.js#L111)). For that reason I would prefer to keep this targeted, and eventually if there's a new case to get a image or something like that, we can update the shim to take those other static assets into account

matthewwalsh0 commented on test/helpers/setup-helper.js:
```diff
@@ -89,10 +91,35 @@ window.document.body.appendChild(popoverContent);
+    if (error.message === 'Only absolute URLs are supported') {
+      const regex = /_locales\/([^/]+)\/messages\.json/gu;
+      const localeRelativePathRequest = url.match(regex);
+      if (localeRelativePathRequest && localeRelativePathRequest.length > 0) {
```
Also here, could we throw again to avoid the nesting?

matthewwalsh0 replied:
Minor, but could this be simplified to `if(localeRelativePathRequest?.length)` ?

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
I may have misunderstood but did the PR description regarding `Test Location` suggest we put the integration tests alongside the components?

seaona replied:
given the nature of the integration tests, I agree it makes more sense to locate the test files under an integration test folder, and if possible, having the same folder structure as we have on e2e (what's currently following this branch). This way, knowing where each test goes will be straight-forward (instead of having to think under which component we should place it, which can be tricky as more than one will come into play), as well as give easily access information for a given feature/wallet area, which are the e2e and integration tests we have. Since those 2 test types complement each other, it can help to see if a given e2e can be ported to integration test and also see the gaps from an integration test and add the corresponding e2e to fill that gap

Gudahtt replied:
For some of these examples, I wonder if we could solve this by refactoring the code such that there was an obvious file to place the test alongside of. For example, in some confirmation refactor discussions we had considered creating one file per type of confirmation. One module per "page" seems like a good way of organizing the codebase. It's good when the code organization reflects how we think about the product.

But the onboarding case you mentioned is a good counter-example. Even if we refactor things to make colocation easier, I can see there being some cases for cross-page integration tests where there would be no obvious home for a test.

Putting the integration tests in their own directory makes sense.

Gudahtt replied:
For test discoverability, it's not perfect but maybe we can have some rule-of-thumb like: page-level components should have integration tests, but any smaller components should be unit-tested. And we can update our `coverageFrom` rules in our unit test config to account for that as well.

cryptotavares replied:
I like the rule-of-thumb for page-level components should have integration tests, everything else should be unit tested 👍 
As for co-locate integration tests with page directories, there are a couple of things that I am not 100% sure:
1. tests that are across multiple pages - though I don't think that this will be common, where would we place the test file? Maybe in the directory of the first page that the test will land? 🤔 
2. By increasing discoverability, will we increase maintenance complexity? i.e. will developers know and follow exactly the guidelines to where create integration tests? I am saying this, because it is fairly common and widespread practice to co-locate unit tests. The same can be said for creating e2e tests within a e2e test directory. If we create a guideline for creating integration tests at the page directory level, will it be followed? Or will we have to enforce it in order to minimise mixing up unit and integration tests? (thus impacting discoverability instead of enhancing it)

Gudahtt replied:
I see what you mean. I think there's a path there somewhere, but we can consider that more carefully in a later PR.

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+  submitRequestToBackground: jest.fn(),
+}));
+
+const mockedBakcgroundConnection = backgroundConnection as jest.Mocked<
```
Minor, but could we also derive this with `jest.mocked(backgroundConnection)`?

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+  onNotification: jest.fn(),
+};
+
+describe('PersonalSign', () => {
```
Minor, but should we add some context such as `Personal Sign Confirmation`?

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+describe('PersonalSign', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    jest.resetAllMocks();
```
My understanding was that `resetAllMocks` functionally includes `clearAllMocks`.

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+    jest.resetAllMocks();
+  });
+
+  afterAll(() => {
```
Is this redundant if we also do it in the `beforeEach`?

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+      ];
+
+    const pendingPersonalSignTime = new Date().getTime();
+    const pendingPersonalSignId = '0050d5b0-c023-11ee-a0cb-3390a510a0ab';
```
Should we define file level constants for these as I think we use them a few times? The same could be argued for any inline raw strings?

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+    const account =
+      mockMetaMaskState.internalAccounts.accounts[
+        mockMetaMaskState.internalAccounts
+          .selectedAccount as 'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3'
```
Why are we casting this to a specific string?

cryptotavares replied:
Because typescript is inferring the types from the mockMetaMaskState.internalAccounts.accounts object keys and not setting it up as just a string type... I wanted to avoid typing the mockMetaMaskState and as it was a test case I didn't spend much time over it.
Maybe we can update that casting to `as keyof typeof mockMetaMaskState.internalAccounts.accounts`? The only other solution that I can think off is typing `mockMetaMaskState`... Do you have any ideas?

matthewwalsh0 commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,259 @@
+    const accountName = account.metadata.name;
+    const pendingPersonalSignTime = new Date().getTime();
+    const pendingPersonalSignId = '0050d5b0-c023-11ee-a0cb-3390a510a0ab';
+    const mockedMetaMaskState = {
```
There's alot of state setup in these tests and it appears lots of it is duplicated. Should we define some constants for this test specific state that we can copy and mutate? Or alternatively a helper function such as `buildState` with property arguments for the state that varies between tests?

matthewwalsh0 commented on test/integration/data/integration-init-state.json:
Is the intent for this to be a base state for all the other integration tests to add to? This feels like alot so are we confident this is the minimum needed for the tests?

Also, could we develop a system similar to fixtures in the E2E tests where we use the builder pattern to create re-usable chunks of state to also help test readability?

I appreciate we can't use the existing fixtures directly since they're setting the raw state, rather than the flattened UI state.

cryptotavares replied:
Ok so two things:
1. yes this is the intent to be the base state for all other tests. I wasn't trying to figure out the minimum. Just reused an initial state from the unit tests. But I was thinking about just getting the state from a freshly installed wallet.
2. Humm I think that we can re-use fixtures actually... but we would need to set up that infrastructure (fixture -> run migrations -> create composableStore -> flatten state). I don't think that it should be part of this PR though. For now I would keep this one simple

Gudahtt replied:
Perhaps we could use the existing e2e fixture builder dynamically create this flatted UI state? We had independently planned to build automation to keep that up-to-date

e.g. something like

```
function getInitegrationinitState() {
  const fixtureState = new FixtureBuilder().build();
  return Object.values(fixtureState).reduce((mergedState, controllerState) => Object.assign(mergedState, controllerState), {});
}
```

No need for any additional automation. That would let us delete this file and generate state dynamically.

cryptotavares replied:
ahah that was similar to my first try, though I was also running the migrations (to ensure that we would always have a latest state version)... but it is not that easy, because the e2e fixtures only have the persisted state. I know that you already know this, but for everyone else, **persisted state might not the same as the controllers state** (it really depends on the controller metadata and what state is persisted). By doing a quick search, all of the above controllers have state that is not persisted:
* SignatureController
* ApprovalController
* KeyringController
* QueuedRequestController
* RateLimitController

I have branch (locally) on top of this one that adds utility functions to get intended integrations test state from an initial UI state (that I got from a freshly installed wallet). The major downsize is that it does not support migrations.. I'll give another spin to try to get the initial state from the fixture builder (maybe by manually adding non persisted state? 🤔 )

Gudahtt replied:
Ohhh right. Hmm. That's tricky.

matthewwalsh0 commented on test/integration/onboarding/wallet-created.test.tsx:
```diff
@@ -0,0 +1,117 @@
+
+    fireEvent.click(getByTestId('onboarding-complete-done'));
+
+    await waitFor(() => {
```
As with E2E tests, I imagine these tests will deal with selectors directly and big assertions. So similar to E2E tests, should we encourage the user of local functions and helper classes to abstract these interactions and complex assertions?

cryptotavares replied:
Not 100% sure we should create those abstractions, at least not from the start.. as the assertions/interactions here are pretty much the same as we would write unit tests.

cryptotavares committed:
chore: split handle relative urls from fetch

This is for improved readability

georgewrmarshall commented:
Regarding UI integration testing, Storybook now provides functionality for [UI integration tests](https://storybook.js.org/docs/writing-tests/interaction-testing). This could allow non-engineering folks to run these tests and encourage creating stories so our UI can be viewed and worked on in isolation. I haven't created any yet, and I wasn't exactly sure how these might fit next to our e2e tests. Testing with production is more stable, but I wanted to provide this for consideration

georgewrmarshall commented:
Storybook now provides functionality for [UI interaction tests](https://storybook.js.org/docs/writing-tests/interaction-testing). Do you see any crossover or consider these a replacement for integration tests? This could allow non-engineering folks to run these tests and encourage creating stories so our UI can be viewed and worked on in isolation. I haven't created any yet, and tbh I wasn't exactly sure how these interaction tests fit in with integration and e2e testing. Testing with production is more stable, but I wanted to provide this for consideration and get your thoughts @cryptotavares

cryptotavares committed:
test: simplify personal sign integration tests

cryptotavares committed:
test: simplify wallet created events integration tests

cryptotavares commented:
> Storybook now provides functionality for [UI interaction tests](https://storybook.js.org/docs/writing-tests/interaction-testing). Do you see any crossover or consider these a replacement for integration tests? This could allow non-engineering folks to run these tests and encourage creating stories so our UI can be viewed and worked on in isolation. I haven't created any yet, and tbh I wasn't exactly sure how these interaction tests fit in with integration and e2e testing. Testing with production is more stable, but I wanted to provide this for consideration and get your thoughts @cryptotavares

@georgewrmarshall Humm I don't think that the interaction tests provided by storybook are a replacement for integration tests. The interaction tests provide the interactivity that is required to write these integration tests, but also a lot of the unit tests. At its core, the integration tests being proposed here, is just a different way to think about testing the UI, while leveraging the same tools that developers are already used to (i.e. jest and testing library).
Now, I think that the question that you are making is, should we start relying on storybook to write our tests? (As it now seems to support all the required capabilities to write integration tests or even e2e tests, through playwright or cypress). That is a broader discussion for which I don't have a good answer as I haven't tried it enough. Going through the docs, it does seem a powerful tool, but we need to also take into account:
* developer experience (as in, how much disruption would it cause? is it going to encourage developers to create more or less tests? would we increase complexity by further fragmenting the testing tools per test type and project/repo? etc etc)
* performance (how fast do those tests run? does it support the same specs as jest dom? maybe it supports even more, as jest does not support scrollability for example)
* scalability/maintenance (how scalable is storybook? not just the tool itself, but in adopting new features. What about mobile? How does it compare with jest and testing library? Does it have increased maintenance costs? etc)
Maybe a good topic to raise awareness in the engineering technical sync meeting.

seaona commented on .circleci/config.yml:
```diff
@@ -1630,6 +1635,23 @@ jobs:
+            - coverage
+      - store_test_results:
+          path: test/test-results/integration/junit.xml
+
```
:information_source:  as info (no change required), in the future it's planned to move the unit test ci jobs to github actions. I think we'll want to do the same with integrations tests?

https://github.com/MetaMask/MetaMask-planning/issues/2449

seaona commented on .circleci/config.yml:
```diff
@@ -221,10 +221,14 @@ workflows:
       - test-unit-jest-development:
           requires:
             - prep-deps
+      - test-integration:
```
do you know what's the `test-integration-main` that appears in the [ci job](https://app.circleci.com/insights/github/MetaMask/metamask-extension/workflows/test_and_release/jobs?branch=cryptotavares%2Fpoc-integration-tests)? :thinking: 

I can't seem to find this change in this branch, but this job seems to only appear in this branch :thinking: 
![Screenshot from 2024-06-25 10-55-57](https://github.com/MetaMask/metamask-extension/assets/54408225/0cb17a41-1c95-4ebd-a184-067c42a65ca8)

cryptotavares replied:
Initially we named it as `test-integration-main` but then thought that it would be better to just call it `test-integration` 😅

seaona replied:
ohhh that explains it! Thank you :laughing:

seaona commented on test/integration/data/integration-init-state.json:
```diff
@@ -0,0 +1,2042 @@
+      "decimals": "18"
+    }
+  ],
+  "transactions": [
```
as an init state, do we want to have already several prepopulated tx's, networks etc.. or would we rather have a more "clean" state, and then add the specific data  on each test that needs it?

cryptotavares replied:
ok so @matthewwalsh0 also hinted at the same. I would rather follow the fixtures approach, but I felt that it would significantly raise complexity for this PR. That being said, I'll come up with a proposal on how to have the fixtures style approach for the integration tests

chloeYue commented:
Hi @cryptotavares, this PR is great initiative that surely opens up exciting new possibilities for testing!  I have two questions:
- The PR description contains a lot of useful information. Do you plan to document this in either the README file or the contributor-doc? This would make it easier to search for and follow the instructions.
- What are the next steps for integration tests?

cryptotavares commented:
> * The PR description contains a lot of useful information. Do you plan to document this in either the README file or the contributor-doc? This would make it easier to search for and follow the instructions.

@chloeYue on some of the initial chats, my understanding was that the platform teams or the wallet framework team would officialise the guidelines in the contributor docs (I am guessing that the recent api testing tool introduced by the wallet api team, should also be referenced there). The PR description that I added here was just to provide further context on the different types of tests, and highlight the importance of integration tests. That being said, I am happy to open a draft PR in the contributor docs with this PR description. I'll defer to @Gudahtt on the best way to proceed with documentation.

> * What are the next steps for integration tests?

Next steps should be for teams to start adding integration tests 😅 We can briefly intro them in the technical engineering sync and highlight them in the slack channels, but then it is up for the teams to start adding tests (the same way that it is up for the different teams to add unit tests or e2e tests). We might want to add an initiative on the QA working group to drive adoption of more integration tests. But I'll defer that to the QA working group and the platform teams

Gudahtt commented:
>my understanding was that the platform teams or the wallet framework team would officialise the guidelines in the contributor docs (I am guessing that the recent api testing tool introduced by the wallet api team, should also be referenced there). The PR description that I added here was just to provide further context on the different types of tests, and highlight the importance of integration tests. That being said, I am happy to open a draft PR in the contributor docs with this PR description

Having these guidelines in contributor-docs would be great! The contributor docs are meant to be maintained by all teams, I don't think we have an expectation of the platform teams being the ones to propose articles or make things official.

Gudahtt commented on ui/pages/confirmations/components/confirm/header/header-info.tsx:
```diff
@@ -200,7 +204,7 @@ const HeaderInfo = ({
               <ConfirmInfoRowCurrency
                 value={balanceToUse ?? 0}
-                dataTestId="header-balance"
+                data-testid="header-info__account-details-modal__account-balance"
```
Nit: This ID seems like it was intended to follow BEM naming conventions (judging by the `__` separator), but it has two "element" parts which is not compliant with BEM.

A BEM name would be something more like

```suggestion
                data-testid="header-info__account-details-modal-balance"
```

or 

```suggestion
                data-testid="confirmation-account-details-modal__account-balance"
```

cryptotavares replied:
I see your point.. I wasn't, intentionally, trying to follow BEM naming conventions. I don't think that there's a convention to the `data-testid` (maybe we should have one). I think I was just splitting by "component hierarchy". Anyways, I'll update them to avoid confusion.

cryptotavares replied:
Also we probably should have some guidelines/conventions on how to name `data-testid`. Otherwise we might inadvertently end up with two elements with the same `data-testid` in the dom, which can lead to unexpected issues when writing tests. Also I find myself heavily relying on the `data-testid`, which seems to be the recommended locator strategy for our e2e tests (according to the contributor docs).
However, only recommends querying by testid if it is not possible to use other queries (except querying dom structure or css classes). Their ordered recommendation for queries can be found [here](https://testing-library.com/docs/queries/about/#priority)

cryptotavares replied:
done 👍

Gudahtt commented on test/helpers/setup-helper.js:
```diff
@@ -89,10 +91,45 @@ window.document.body.appendChild(popoverContent);
 const { default: fetch, Headers, Request, Response } = require('node-fetch');
 
-Object.assign(window, { fetch, Headers, Request, Response });
+const handleRelativePathRequest = async (url, localeRelativePathRequest) => {
```
Interesting solution! As an alternative to this, perhaps we can update `ui/contexts/i18n.js` to import the `en` localized messages rather than fetching them at runtime.  I see that we're already doing this in `shared/modules/i18n.js`. This would remove the need for shimming fetch I think.

cryptotavares replied:
Ooooh this is interesting... We can definitely do that. But there's two things:
1. by importing it we are now including it in the bundled js. Currently, the `en` translations are included in the common.js bundle. They are also included as static assets in the `_locales` dir at the root (where the static.js build script is placing them). Probably the impact on the bundle size is neglectable  (en messages is about 220kb), but still we are duplicating it.
2. _locales are within the `app` directory. We probably want to move it from there, to avoid having the `ui` importing `app` files directly.
3. there's actually a bug on the ui code 😅 

So the [setupLocale](https://github.com/MetaMask/metamask-extension/blob/develop/ui/index.js#L88-L90) method that is run when initialising the UI to set the current locale messages (and the en fallback) in the redux store, always returns an empty object (the [fetchLocale](https://github.com/MetaMask/metamask-extension/blob/develop/shared/modules/i18n.ts#L89-L101) is wrapped on a try/catch that returns an empty object in case there's an error). This means if we fail to fetch/load the local files, the redux store will have this initial state:
```
    localeMessages: {
      ...
      current: {},
      en: {},
    },
```
The problem is that the [UI i18n context](https://github.com/MetaMask/metamask-extension/blob/develop/ui/contexts/i18n.js#L14-L22) and the [getMessage](https://github.com/MetaMask/metamask-extension/blob/develop/ui/helpers/utils/i18n-helper.tsx#L19) method does not handle well the empty object. The getMessage relies on the shared version of it which handles falsy localMessages but not an empty object. In that case it will throw a [missingKeyError](https://github.com/MetaMask/metamask-extension/blob/develop/shared/modules/i18n.ts#L63-L72)
.

cryptotavares replied:
Also the [setupLocale](https://github.com/MetaMask/metamask-extension/blob/develop/ui/index.js#L88-L90) method seems buggy as the if clause to load the current locale relative time format and date will always be true - https://github.com/MetaMask/metamask-extension/blob/develop/shared/lib/error-utils.js#L13 🤔

cryptotavares replied:
Update to the above.. the `missingKeyError` will only throw if we are running the app in test mode (process.env.IN_TEST). That is why the tests were failing. The Fallback seems to be applied if we are not running in test mode.
Still is there a reason to throw when running in test? 🤔

chloeYue replied:
hm...This is interesting. I'm not aware of any specific reason for the decision to throw missingKeyError only during testing 🤔, It might be worth discussing whether this behavior is necessary or should we reconsider the approach.

Gudahtt commented on test/integration/config/env.js:
```diff
@@ -0,0 +1,3 @@
+require('../../env');
+
+process.env.ENABLE_CONFIRMATION_REDESIGN="true"
```
Should we be enabling this globally? Presumably we'd only want this for integration tests of the confirmation redesign

cryptotavares replied:
That is old code that stayed behind. Redesigned signature pages are no longer behind that env variable. I am removing it from the test env.

cryptotavares replied:
done 👍

Gudahtt commented on test/integration/config/setup.js:
```diff
@@ -0,0 +1,10 @@
+require('../../helpers/setup-helper');
+
+window.SVGPathElement = window.SVGPathElement || { prototype: {} };
+global.indexedDB = {};
```
Nit: Much of the contents of this file are identical to `test/setup.js`. I wonder if the common steps should be moved into `setup-helper.js`? e.g. reserve this file for integration-test-specific setup, and `test/setup.js` for unit test specific setup.

Gudahtt replied:
I suppose the same suggestion applies to `setupAfter` as well, except that there is no shared module for that yet.

cryptotavares replied:
done 👍

Gudahtt commented on ui/index.js:
```diff
@@ -79,7 +79,11 @@ export default function launchMetamaskUi(opts, cb) {
 }
 
-async function startApp(metamaskState, backgroundConnection, opts) {
+export async function setupInitialStore(
```
Nit: A JSDoc comment here would be greatly appreciated

Gudahtt commented on test/lib/render-helpers.js:
```diff
@@ -133,3 +135,26 @@ export function renderWithUserEvent(jsx) {
   };
 }
+
+export async function integrationTestRender(extendedRenderOptions) {
```
Nit: A JSDoc comment would be greatly appreciated here as well

Gudahtt commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,190 @@
+    const account =
+      mockMetaMaskState.internalAccounts.accounts[
+        mockMetaMaskState.internalAccounts
+          .selectedAccount as keyof typeof mockMetaMaskState.internalAccounts.accounts
```
Hmm, it looks like the type error you're getting past with this cast is caused by TypeScript interpreting the initial state as a constant. Ideally we'd widen that type to reflect the actual MetaMask state shape, but we don't have a type for that yet. Hopefully this is something we can return to soon.

Gudahtt commented on test/integration/confirmations/signatures/personalSign.test.tsx:
```diff
@@ -0,0 +1,190 @@
+
+    fireEvent.click(getByTestId('header-info__account-details-button'));
+
+    await waitFor(() => {
```
Nit: I haven't seen so many assertions in a `waitFor` callback before, is this typical? i.e. would you expect one of these elements to be found before the others had loaded?

I would expect the metric event might be slightly slower to show up, so that makes sense as an independent thing that we're waiting for. Though that one is a bit strange too, in that if the wrong metric event appeared first, we'd keep running failing assertions on it in a loop until we hit the timeout.

Maybe it'd make more sense to stage this as:
* Wait for account details modal
* Assert contents
* Wait for _some_ metric event to be captured
* Assert contents

Then we wouldn't have any redundant locators, and we'd fail faster if the wrong metric event appears first.

Gudahtt replied:
I see that the same comment applies to a few other large `waitFor` callbacks in the other integration test module as well

cryptotavares replied:
That makes sense. Done 👍

Gudahtt commented on test/integration/onboarding/wallet-created.test.tsx:
```diff
@@ -0,0 +1,109 @@
+  submitRequestToBackground: jest.fn(),
+}));
+
+const mockedBakcgroundConnection = jest.mocked(backgroundConnection);
```
Nit: Typo
```suggestion
const mockedBackgroundConnection = jest.mocked(backgroundConnection);
```

Gudahtt reviewed (dismissed):
Looks great! Left a bunch of suggestions but nothing blocking.

cryptotavares commented:
> > my understanding was that the platform teams or the wallet framework team would officialise the guidelines in the contributor docs (I am guessing that the recent api testing tool introduced by the wallet api team, should also be referenced there). The PR description that I added here was just to provide further context on the different types of tests, and highlight the importance of integration tests. That being said, I am happy to open a draft PR in the contributor docs with this PR description
> 
> Having these guidelines in contributor-docs would be great! The contributor docs are meant to be maintained by all teams, I don't think we have an expectation of the platform teams being the ones to propose articles or make things official.

I'll open up a draft PR with these guidelines 👍

cryptotavares committed:
fix: data-testid naming

The idea was to have a more structured
testid naming. We are following a BEM
convention here.

cryptotavares committed:
fix: remove redesing env from integration tests

cryptotavares committed:
fix: typo

cryptotavares committed:
chore: add jsdoc

Added jsdoc to integration render helper
and setup initial store methods

cryptotavares committed:
chore: improve waitFor usage

This should avoid redundant locators and allow
us to fail faster instead of keeping
running failing assertions in loop

cryptotavares committed:
chore: move common jest setup test setup helper

cryptotavares committed:
chore: move common jest setup after test setup after helper

Gudahtt reviewed (dismissed):
LGTM!

Gudahtt commented:
Oh, another conflict, I can re-review when that is resolved

cryptotavares committed:
chore: add integration tests github action

cryptotavares committed:
fix: confirmations data test id

matthewwalsh0 commented on .github/workflows/run-integration-tests.yml:
```diff
@@ -0,0 +1,18 @@
+    types: [opened,reopened,synchronize]
+
+jobs:
+  test-unit-jest:
```
Minor, should we rename this to `test-intergration-jest`?

cryptotavares replied:
🤦‍♂️

matthewwalsh0 reviewed (dismissed):
❤️

Gudahtt commented on .github/workflows/run-integration-tests.yml:
```diff
@@ -0,0 +1,18 @@
+name: Run integration tests
+
+on:
+  pull_request:
```
We should run this on commits to `develop` and `master` as well

cryptotavares replied:
Should we also run them on release branches?

Gudahtt replied:
Release branches should already have PRs associated with them, so the other trigger should work fine there. `master` and `develop` are the only two long-running branches that we have without PRs

pedronfigueiredo reviewed (dismissed):
LGTM!

cryptotavares committed:
fix: integration test gh action

Fix job name typo and run
action on push to develop and master

Gudahtt reviewed (dismissed):
LGTM!

cryptotavares committed:
fix: shim fetch only for integration tests

Shimming it for both unit and integration
was having almost 80% increase in execution
time for unit tests.

cryptotavares commented:
> > > my understanding was that the platform teams or the wallet framework team would officialise the guidelines in the contributor docs (I am guessing that the recent api testing tool introduced by the wallet api team, should also be referenced there). The PR description that I added here was just to provide further context on the different types of tests, and highlight the importance of integration tests. That being said, I am happy to open a draft PR in the contributor docs with this PR description
> > 
> > 
> > Having these guidelines in contributor-docs would be great! The contributor docs are meant to be maintained by all teams, I don't think we have an expectation of the platform teams being the ones to propose articles or make things official.
> 
> @Gudahtt I'll open up a draft PR with these guidelines 👍

Contributor docs PR - https://github.com/MetaMask/contributor-docs/pull/99
